### PR TITLE
[IAM]: Issue 202: Support aws_iam_policy_attachment resources

### PR DESCRIPTION
**Pull request by stormbeta**
This adds support for attaching managed policies to users, roles, and groups.

I chose to refer to users, groups, or roles by direct name rather than using the lookup syntax because the latter doesn't work if any of those resources aren't also managed by terraform.

### CODE REVIEW

File: lib/terraforming/resource/iam_attached_policies.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            }
+          }
+
+          resources
+        end
+      end
+
+      private
+
+      def iam_policies_map()
```
Comment (dtan4) [inline lib/terraforming/resource/iam_attached_policies.rb]: I think `iam_policy_attachments` is better.

Comment (dtan4) [inline lib/terraforming/resource/iam_attached_policies.rb]: Please remove `()` if no argument is required for the method.

File: lib/terraforming/resource/iam_attached_policies.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        policies = Hash.new
+        [:user, :group, :role].each do |type|
+          list = @client.send("list_#{type}s")["#{type}s"]
+          list.each do |resource|
+            resource_name = resource["#{type}_name"]
+            attached = @client.send("list_attached_#{type}_policies", :"#{type}_name" => resource_name).attached_policies
+            attached.each do |policy|
+              unless policies.key? policy.policy_name then
```
Comment (dtan4) [inline lib/terraforming/resource/iam_attached_policies.rb]: Please remove `then`.

File: lib/terraforming/template/tf/iam_attached_policies.erb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<% iam_policies_map().each do |name, policy| -%>
+resource "aws_iam_policy_attachment" "<%= name %>-attachments" {
+    name       = "<%= name %>-attach"
+    users      = [<%= policy[:user].collect{|x| "\"#{x}\""}.join(',') %>]
```
Comment (dtan4) [inline lib/terraforming/template/tf/iam_attached_policies.erb]: Use `user` instead of `x`. (following `role` and `group` are the same.)

**Comment by dtan4** [comment]
Other resources in Terraforming have class names which are equivalent to Terraform resource name.
According to this rule, this resource class should be named `IAMPolicyAttachment`.

[AWS: aws_iam_policy_attachment - Terraform by HashiCorp](https://www.terraform.io/docs/providers/aws/r/iam_policy_attachment.html)

**Comment by stormbeta** [comment]
Thanks for the feedback! I've updated it to match your suggestions.

File: lib/terraforming/cli.rb [hunk 4]
```diff
@@ -51,6 +51,11 @@ def elb
       execute(Terraforming::Resource::ELB, options)
     end
 
+    desc "iampa", "IAM Policy Attachment"
+    def iampa
```
Comment (dtan4) [inline lib/terraforming/cli.rb]: Please move this task next to `iamp`.
All tasks are sorted by ASCII order.

File: lib/terraforming/resource/iam_policy_attachment.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            }
+          }
+
+          resources
+        end
+      end
+
+      private
+
+      def iam_policies_map
```
Comment (dtan4) [inline lib/terraforming/resource/iam_policy_attachment.rb]: Hmm, this is still remained... Please rename this to `iam_policy_attachments`.

File: lib/terraforming/resource/iam_policy_attachment.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                    role:  Set.new,
+                }
+              end
+
+              policy_entry = policies[policy.policy_name]
+              policy_entry[type].add(resource_name)
+            end
+          end
+        end
+        return policies
```
Comment (dtan4) [inline lib/terraforming/resource/iam_policy_attachment.rb]: This `return` can be omitted.

**Comment by trompx** [comment]
Hello,

I have multiple users, multiples policies and policies attached to users. `terraformin iamu/iamp` return correct data.
However, when I use `terraforming iamup` I get nothing. On terraform (https://www.terraform.io/docs/providers/aws/r/iam_user_policy.html), user policy is equivalent to attached policies. Is there any difference with user policies ?
Is this going to be merged in the next release ? 

Thanks

**Comment by dtan4** [comment]
IAM Policy Attachment support was implemented at https://github.com/dtan4/terraforming/pull/225.
Thank you for sending Pull Request, and please use the latest Terraforming.

**Pull request closed without merging**